src/search.cpp

```diff
@@ -211,18 +200,70 @@ void Search::clear() {
   Tablebases::init(Options["SyzygyPath"]); // Free mapped files
 }
 
+/// MainThread::perft() divide the perft search evenly through Threads
+void MainThread::perft() {
+
+  StateInfo st;
+  std::string posNow = rootPos.fen();
+  std::vector<perftPos> perftRoot, newPerftRoot;
+
+  // Step 1: Create new root positions that will be distributed between threads
+  for (const auto& m : MoveList<LEGAL>(rootPos)) {
+      rootPos.do_move(m, st);
+      perftSum[perftRoot.size()].store(0, std::memory_order_release);
+      perftRoot.push_back({rootPos.fen(), Depth(Limits.perft - 1), (int)perftRoot.size()});
+      rootPos.undo_move(m);
+  }
+
+  // Step 2: Extend root positions if it's not enough to distribute evenly between threads
+  while (perftRoot.size() && perftRoot[0].d > 3 && 8 * Threads.size() > perftRoot.size()){
+      newPerftRoot.clear();
+      for (const auto& pm : perftRoot){
+          rootPos.set(pm.fen, rootPos.is_chess960(), rootPos.state(), this);
+
+          for (const auto& m : MoveList<LEGAL>(rootPos)){
+              rootPos.do_move(m, st);
+              newPerftRoot.push_back({rootPos.fen(), pm.d - 1, pm.ID});
+              rootPos.undo_move(m);
+          }
+      }
+      perftRoot = newPerftRoot;
+  }
```

Comment:
Are the terminal nodes here counted correctly?

Reply:
They are not expanded, because of `perftRoot[0].d > 3`. And they are handled inside `Thread::search()`.